demo/demo.ts

```diff
@@ -3,60 +3,37 @@ import {EditorView} from "../view/src/"
 import {keymap} from "../keymap/src/keymap"
 import {history, redo, redoSelection, undo, undoSelection} from "../history/src/history"
 import {gutter} from "../gutter/src/index"
-import {baseKeymap} from "../commands/src/commands"
+import {baseKeymap, indentSelection} from "../commands/src/commands"
 import {legacyMode} from "../legacy-modes/src/index"
 import {matchBrackets} from "../matchbrackets/src/matchbrackets"
 import javascript from "../legacy-modes/src/javascript"
 import {specialChars} from "../special-chars/src/special-chars"
 import {multipleSelections} from "../multiple-selections/src/multiple-selections"
 
-let mode = legacyMode(javascript({indentUnit: 2}, {}))
-
-// FIXME these should move to commands and access the indentation
-// feature through some kind of generic mechanism that allows plugins
-// to advertise that they can do indentation
-function crudeInsertNewlineAndIndent({state, dispatch}: EditorView): boolean {
-  let indentation = (mode as any).indentation(state, state.selection.primary.from)
-  if (indentation > -1)
-    dispatch(state.transaction.replaceSelection("\n" + " ".repeat(indentation)).scrollIntoView())
-  return true
-}
-function crudeIndentLine({state, dispatch}: EditorView): boolean {
-  let cursor = state.selection.primary.head // FIXME doesn't indent multiple lines
-  let line = state.doc.lineAt(cursor), text = line.slice(0, 100)
-  let space = /^ */.exec(text)[0].length // FIXME doesn't handle tabs
-  let indentation = (mode as any).indentation(state, line.start)
-  if (indentation == -1) indentation = space
-  let tr = state.transaction.replace(line.start, line.start + space, " ".repeat(indentation)).scrollIntoView()
-  if (cursor <= line.start + space)
-    tr = tr.setSelection(EditorSelection.single(line.start + indentation))
-  dispatch(tr)
-  return true
-}
+let mode = legacyMode.use({mode: javascript({indentUnit: 2}, {}) as any})
 
 let isMac = /Mac/.test(navigator.platform)
 let state = EditorState.create({doc: `"use strict";
 const {readFile} = require("fs");
 
 readFile("package.json", "utf8", (err, data) => {
   console.log(data);
-});`, plugins: [
-  gutter(),
-  history(),
-  specialChars({}),
-  multipleSelections(),
+});`, extensions: [
+  gutter.use(),
```

Comment:
I agree.

Reply:
Also agree, and I actually seriously looked into doing the scary things required to make a class derive from `Function`, but decided it was too ugly and unexpected (you have to initialize a function and then patch its prototype, the types will be strange, etc)

Reply:
Interesting how the limitations of TypeScript shape the public facing API.

It reminds me of the D3 Types. Many D3 "classes" are functions that have properties that are also functions. The types for `d3-axis` may be interesting to study  https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3-axis/index.d.ts#L64

```js
const myAxis = d3.axisLeft(scale); // Constructor
myAxis(selection); // Instance invoked as a function
```

Perhaps a public facing wrapper (interface?) around the original class would enable the nicer API.

Reply:
TypeScript can type such objects just fine. The important consideration is using concepts that people are familiar with, and not writing confusing code.

Reply:
I took another stab at this, since it _really_ makes code look nicer, but mixing call signatures with classes is still a complete mess in JS (and even more so in TS). So if we do this, we'd have to make `Behavior` and `Extension` types, not classes, which means we don't get static methods and instanceof or any other class-related functionality.

Reply:
Patch bbe5384c85 does this. It moves the behavior accessors to the state class, so that, as far as client code is concerned, behaviors and extensions are just functions. This does involve some kludgery, but it stays within the module, and I think client code won't be affected too much.

Reply:
Excellent! Very happy to see this. Thank you for your effort here.

Reply:
Oh no I really should read notifications before I spent hours on something you're gonna do quicker than me anyway :/

Reply:
Oh, right, don't build on this code—I'm still working on it and we'll just make conflicting changes!

Reply:
How far did you get? Did you hit on any improvements compared to my approach?

Reply:
I think I'm nearly done, but I think our approaches are roughly the same.